site/bps/BP-38-bypass-journal-ledger.md

```diff
@@ -0,0 +1,90 @@
+This proposal is aimed at providing bypass journal ledger, this feature includes these parts work:
+ - add new write flag `BYPASS_JOURNAL` to existing protocol
+ - impl the newly write flag at the client side and server side
+ 
+The details of changes are listed in Section “[Proposed Changes](#proposed-changes)”.
+ 
+### Public Interfaces
+
+This feature will introduce new 'WRITE_FLAG' to enable single write storage impl.
+Like `DEFERRED_SYNC`, when we construct `WriteHandle`, we can pass `BYPASS_JOURNAL`, then every `AddRequest` will carry this flag.
+As for monitoring, we can add metrics to record the main time interval during the io path.
+The detailed compatibility related stuff is in Section “[CDMP](#compatibility-deprecation-and-migration-plan)”.
+
+### Proposed Changes
+
+While the main approach is based on [`WRITE_FLAG` impl](https://github.com/apache/bookkeeper/pull/742),
+the actual implementation is more tricky. In my opinion, there are three different solution:
+
+1. Relax LAC protocol
+
+    Modify server side code mostly and don't change legerHandle's LAC advance logic, if the write flag is `BYPASS_JOURNAL`, after write to `LegerStorage`(the data maybe in the memTable, or the buffer of File, or the os cache),
+bookie return result to the client directly.
+This impl is like [disable syncData](https://github.com/apache/bookkeeper/issues/753), once all the replica fails, the BK cluster can't recovery from it.
+Note, the user shall know the weak durability for his use case when using this impl.
+
+2. Extend `Force` API
+
+    Like `DEFERRED_SYNC`, the normal 'bypass-journal' operation don't advance LAC on the client.
+Only advance LAC using `force` api. To support this new 'bypass-journal' option, we need theses changes:
+
+    - Add persistent callback to LedgerStorage
+    
+        Maintain non-persistent entry list(necessary for out of order entry arriving) and `maxPersistentEntryId` on `LedgerDescriptor`
+If the entry is 'bypass-journal', the callback for this entry is not null, and it will receives notification after persistent through legerStorage.
+    
+    - Extend the force request
+        
+        If the force request contains bypass journal option, get `maxPersistentEntryId` from `LedgerDescriptor`
+ or impl force method on `LedgerDescriptor`, which force the entry to persistent device through `LedgerStorage`.
+
+    In addition to these changes, we should consider who is responsible for the `force` execution? How often to schedule 'force'?
+
+3. Add NonPersistentLAC semantic to existing LAC protocol
+
+    To not violate LAC semantics like method 1, we can add `NonPersistentLAC` semantic, this includes these changes:
+
+    - Add persistent callback to LedgerStorage
+
+        Maintain non-persistent entry list and `maxPersistentEntryId` on `LedgerDescriptor`.
```

Comment:
I tried to have a 'maxPersistentEntryId' while working on DEFERRED_SYNC but actually it was not possible because if you use "LedgerHandleAdv" entries won't get to the bookie with a specific order, so you will have to keep track of holes.

Reply:
you're right, thanks